### ISSUE TITLE
Support strict mode

### ISSUE THREAD

**Issue by wkeese** [SitePen/dstore#244]
I've started to get exceptions using `Cache` that

> strict mode inherited() requires the caller function to be passed before arguments

even though the dstore code itself doesn't have "use strict".

I guess the code needs to be updated to use `this.inherited(caller, arguments`) syntax.

**Comment by arndbeissner** [comment]
Yes, you often need to do that. This can for example happen when you use TypeScript and switch from old style Dojo modules to ES6 modules (or even TS classes). In that case, tsc will start compiling to strict mode and you will see the above errors when your code this.inherited without the function as first arg. The good news is you don't need to patch Dojo, it's sufficient (at least in my experience) to switch the this.inherited syntax in your code.

**Comment by wkeese** [comment]
That's what I would have expected too, but my code doesn't have any `this.inherited()` calls.  The stack trace for the exception is:

```
Notifier.js:41 decor/Notifier: exception Error: declare ccore/store/SortingFilteringPagingCache: strict mode inherited() requires the caller function to be passed before arguments
    at err (declare.js:15)
    at TMP.inherited (declare.js:116)
    at TMP._fetch (Cache.js:76)
    at TMP.fetchRange (SortingFilteringPagingCache.js:1)
```

The key thing here (I think) is that it's complaining about the `this.inherited()` call in dstore's Cache.js.

Maybe it's because Cache.js is a weird case, it's doing funky stuff with `lang.delegate()` and `dojo.safeMixin()` rather than just a straight inheritance

**Comment by arndbeissner** [comment]
I checked, and we no longer use dstore/Cache. I did a quick test with adding cache to a Memory, Trackable combination in a strict mode environment, and that produced no immediate problem, so I guess some context is necessary to reproduce the issue. You may need to attach some demo code for people to look into it.

**Comment by wkeese** [comment]
OK, so my original code was:

```
wrapper = Cache.create(store, {
	...
	// Override Cache#fetchRange() to first populate the cache (if it's not already populated,
	// and there's not an in-progress fetch to populate it.)
	fetchRange: function () {
		if (!this.isAvailableInCache()) {
			this.fetch();
		}
		return this._fetch(arguments, true);
	}
});
```

So presumably the problem was that the `arguments` variable passed to `this._fetch()` comes from a strict environment, breaking the `this.inherited()` call in dstore/Cache.js, even though Cache.js itself isn't strict.


There was also another problem with that code, that `this` would sometimes refer to a child collection (`wrapper.filter(...).fetchRange()`), whereas I wanted to reference the original wrapper created by `Cache.create()`.   So I ended up with:

```
// When fetchRange() is called on the wrapper, or one of the wrapper's child collections, first populate
// the wrapper's cache (if it's not already populated, and there's not an in-progress fetch to populate it).
advise.before(wrapper, "fetchRange", function () {
	// Use "wrapper" rather than "this", so we update the wrapper rather than a child collection.
	if (!wrapper.isAvailableInCache()) {
		wrapper.fetch();
	}
});
```

Long story short, at this point I guess there's no reason to update dstore , and we can close this ticket.   Thanks for looking at it with me.

**Comment by wkeese** [comment]
PS: I guess a reason to update it would be so people could freely make `Cache` subclasses etc. (in strict mode) that could call `_fetch()`.  It would need a signature like `_fetch(caller, arguments, isRange)`.   But I don't personally need that now.
